File: 백준/Silver/4963. 섬의 개수/섬의 개수.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] dx = {1, 0, -1, 0, 1, -1, -1, 1}; //대각선까지
    static int[] dy = {0, 1, 0, -1, 1, -1, 1, -1};
    static int[][] graph;
    static boolean[][] visited;
    static int w, h;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        while(true){ 
            st = new StringTokenizer(br.readLine());
            w = Integer.parseInt(st.nextToken());
            h = Integer.parseInt(st.nextToken());
            if(w==0 && h==0){//0 0 입력되면 종료
                break;
            }
            graph = new int[h][w];
            visited = new boolean[h][w];
            int count = 0;
            for(int i=0; i<h; i++){
                st = new StringTokenizer(br.readLine());
                for(int j=0; j<w; j++){
                    graph[i][j] = Integer.parseInt(st.nextToken());
                }
            }
            for(int i=0; i<graph.length; i++){
                for(int j=0; j<graph[i].length; j++){
                    if(graph[i][j]==1 && !visited[i][j]){
                        visited[i][j] = true;
                        bfs(i, j);
                        count++;
                    }
                }
            }
            sb.append(count).append("\n");
        }
        br.close();
        System.out.println(sb.toString());
    }
    static void bfs(int x, int y){
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        visited[x][y] = true;

        while(!q.isEmpty()){
            int[] cur = q.poll();
            for(int i=0; i<8; i++){
                int nx = cur[0]+dx[i];
                int ny = cur[1]+dy[i];

                if(nx<0 || nx>=h || ny<0 || ny>=w) continue;
                if(graph[nx][ny] == 0 || visited[nx][ny]) continue;
                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny});
            }
        }
    }
}

File: 백준/Silver/4963. 섬의 개수/섬의 개수.py
from collections import deque
import sys
input = sys.stdin.readline
dx = [1, 0, -1, 0, 1, 1, -1, -1]
dy = [0, 1, 0, -1, 1, -1, 1, -1]

def bfs(x, y):
    visited[x][y] = True
    q = deque([(x, y)])
    while q:
        curx, cury = q.popleft()
        for i in range(8):
            nx, ny = curx+dx[i], cury+dy[i] 
            if 0<=nx<h and 0<=ny<w:
                if not visited[nx][ny] and graph[nx][ny]==1:
                    visited[nx][ny] = True
                    q.append((nx, ny))
    
while True:
    w, h = map(int, input().split())
    if w==0 and h==0: break
    graph = []
    visited = [[False]*w for _ in range(h)]
    count = 0
    for _ in range(h):
        graph.append(list(map(int, input().split())))
    
    for x in range(h):
        for y in range(w):
            if graph[x][y]==1 and not visited[x][y]:
                bfs(x, y)
                count +=1
    print(count)